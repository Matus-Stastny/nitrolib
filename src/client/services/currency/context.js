// @flow
import * as React from "react";

import { currencyDefault } from "client/records/Currency";

// $FlowIssue
export const { Consumer, Provider } = React.createContext(currencyDefault);
