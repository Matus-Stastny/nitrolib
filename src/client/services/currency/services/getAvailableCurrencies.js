// @flow strict
import * as R from "ramda";

import type { Currencies } from "client/records/Currency";

const getAvailableCurrencies = (currencies: Currencies, affiliate: string): Currencies =>
  R.filter(
    R.compose(
      R.either(
        R.complement(R.is(Array)),
        R.any(
          // eslint-disable-next-line no-underscore-dangle
          R.contains(R.__, R.toLower(affiliate)),
        ),
      ),
      R.prop("enabledOnAffilId"),
    ),
    currencies,
  );

export default getAvailableCurrencies;
