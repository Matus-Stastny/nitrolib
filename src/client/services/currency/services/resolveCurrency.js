// @flow strict
import * as R from "ramda";

import { currencyDefault } from "client/records/Currency";
import type { Currency, Currencies } from "client/records/Currency";

function resolveCurrency(all: Currencies, available: Currencies, candidate: string): Currency {
  if (available[candidate]) {
    return available[candidate];
  }

  if (all[candidate] && available[all[candidate].fallback]) {
    return available[all[candidate].fallback];
  }

  if (available[currencyDefault.id]) {
    return available[currencyDefault.id];
  }

  return available[R.head(R.sortBy(R.identity, R.keys(available)))];
}

export default resolveCurrency;
