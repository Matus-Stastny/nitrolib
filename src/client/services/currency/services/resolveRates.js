// @flow strict
import * as R from "ramda";

import type { Currencies } from "client/records/Currency";

type Data = {|
  currencies: {|
    edges: Array<{|
      node: {|
        code: string,
        rate: number,
      |},
    |}>,
  |},
|};

const resolveRates = (data: Data, currencies: Currencies) => {
  const ratesMap = data.currencies.edges.reduce(
    (acc, item) => R.assoc(item.node.code, item.node.rate, acc),
    {},
  );

  return R.map(
    currency => R.assoc("rate", R.propOr(currency.rate, currency.id, ratesMap), currency),
    currencies,
  );
};

export default resolveRates;
