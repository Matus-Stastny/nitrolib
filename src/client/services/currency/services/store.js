// @flow strict
import jsCookie from "js-cookie";

import { CURRENCY as COOKIE_CURRENCY } from "client/consts/cookies";

export function getValue() {
  return jsCookie.get(COOKIE_CURRENCY);
}

export function saveValue(newValue: string) {
  jsCookie.set(COOKIE_CURRENCY, newValue);
}
