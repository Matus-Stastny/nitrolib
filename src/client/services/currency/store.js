// @flow strict
import jsCookie from "js-cookie";

import { CURRENCY } from "client/consts/cookies";

export function getValue() {
  return jsCookie.get(CURRENCY);
}

export function saveValue(newValue: string) {
  jsCookie.set(CURRENCY, newValue);
}
