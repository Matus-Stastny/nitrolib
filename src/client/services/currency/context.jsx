// @flow strict
/* eslint-disable react/no-unused-prop-types, react/no-unused-state */
import * as React from "react";
import idx from "idx";

import { currencyDefault } from "client/public/records/Currency";
import type { Currency, Currencies } from "client/public/records/Currency";
import type { Countries } from "client/public/records/Country";
import filterCurrencies from "./services/filterCurrencies";
import resolveCurrency from "./services/resolveCurrency";
import * as store from "./services/store";
import getGeoCountryCall from "./services/getGeoCountry";
import getCurrenciesCall from "./services/getCurrencies";
import getCandidate from "./services/getCandidate";

type Props = {|
  whitelist: string[],
  countries: Countries,
  affiliate: string,
  ip: string,
  initialCurrency: string,
  langCurrency: string,
  children: React.Node,
  // DI
  getValue: typeof store.getValue,
  saveValue: typeof store.saveValue,
  getCurrencies: typeof getCurrenciesCall,
  getGeoCountry: typeof getGeoCountryCall,
|};

type State = {|
  currency: ?Currency,
  all: Currencies,
  available: Currencies,
  country: string,
|};

type Context = {|
  currency: Currency,
  available: Currencies,
  setCurrency: (code: string) => void,
|};

const { Consumer, Provider } = React.createContext(
  ({
    currency: currencyDefault,
    available: { eur: currencyDefault },
    setCurrency: () => {},
  }: Context),
);

// FIXME try to rewrite to Relay
// would require too much 'Currencies' type refactoring, so not done immediately
export class CurrencyProvider extends React.PureComponent<Props, State> {
  static defaultProps = {
    getValue: store.getValue,
    saveValue: store.saveValue,
    getCurrencies: getCurrenciesCall,
    getGeoCountry: getGeoCountryCall,
  };

  state = {
    currency: null,
    all: {},
    available: {},
    country: "",
  };

  componentDidMount() {
    this.loadData();
  }

  componentDidUpdate(props: Props, state: State) {
    if (!state.all) {
      this.loadData();
    }
  }

  static getDerivedStateFromProps(props: Props, state: State) {
    if (state.currency !== null) {
      return null;
    }

    if (!state.all || !state.country) {
      return null;
    }

    const candidate = getCandidate({
      initial: props.initialCurrency,
      country: idx(props.countries, _ => _[state.country].currency) || "",
      lang: props.langCurrency,
    });

    const available = filterCurrencies(props.affiliate, props.whitelist, state.all);

    return {
      currency: resolveCurrency(state.all, available, candidate),
      available,
    };
  }

  setCurrency = (code: string) => {
    const { available } = this.state;
    const { saveValue } = this.props;

    const currency = available[code];
    if (currency) {
      this.setState({ currency });
      saveValue(code);
    }
  };

  async loadData() {
    const { getCurrencies, getGeoCountry, ip } = this.props;

    const [all, country] = await Promise.all([getCurrencies(), getGeoCountry(ip)]);

    this.setState({ all, country });
  }

  render() {
    const { currency, available } = this.state;
    const { children } = this.props;

    if (!currency) {
      return children;
    }

    return (
      <Provider
        value={{
          currency,
          available,
          setCurrency: this.setCurrency,
        }}
      >
        {children}
      </Provider>
    );
  }
}

export { Consumer, CurrencyProvider as Provider };
