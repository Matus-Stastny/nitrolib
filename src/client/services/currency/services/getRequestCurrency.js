// @flow strict
import queryString from "query-string";

import { CURRENCY as URL_CURRENCY } from "client/consts/url";
import * as store from "./store";

const getRequestCurrency = (search: string) =>
  queryString.parse(search)[URL_CURRENCY] || store.getValue() || null;

export default getRequestCurrency;
