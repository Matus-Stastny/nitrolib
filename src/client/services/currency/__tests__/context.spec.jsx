// @flow strict
import * as React from "react";
import { MemoryRouter } from "react-router-dom";
import { shallow } from "enzyme";

import fetchGeoData from "../services/CurrencyGeo";
import fetchRatesData from "../services/CurrencyRates";
import getAvailableCurrencies from "../services/getAvailableCurrencies";
import getRequestCurrency from "../services/getRequestCurrency";
import resolveCurrency from "../services/resolveCurrency";
import resolveRates from "../services/resolveRates";
import * as store from "../services/store";
import { Provider } from "../context";

jest.mock("../services/CurrencyGeo");
jest.mock("../services/CurrencyRates");
jest.mock("../services/getAvailableCurrencies");
jest.mock("../services/getRequestCurrency");
jest.mock("../services/resolveCurrency");
jest.mock("../services/resolveRates");
jest.mock("../services/store");

const currencies = {
  eur: {
    id: "eur",
    name: "Euro",
    format: "__price__ €",
    uncertainFormat: false,
    round: "2",
    enabledOnAffilId: "",
    fallback: "",
    rate: 1,
  },
  gbp: {
    id: "gbp",
    name: "British Pound Sterling",
    format: "£__price__",
    uncertainFormat: false,
    round: "2",
    enabledOnAffilId: ["UK"],
    fallback: "",
    rate: 1.14355,
  },
  czk: {
    id: "czk",
    name: "Czech Republic Koruna",
    format: "__price__ Kč",
    uncertainFormat: false,
    round: "0",
    enabledOnAffilId: "",
    fallback: "",
    rate: 0.0387945,
  },
};

const countries = {
  gb: {
    id: "gb",
    continent: "eu",
    currency: "gbp",
    EN: "United Kingdom",
  },
  cz: {
    id: "cz",
    continent: "eu",
    currency: "czk",
    EN: "Czech Republic",
  },
};

describe("#Currency Provider", () => {
  beforeEach(() => {
    // $FlowIssue
    fetchGeoData.mockReturnValue(Promise.resolve(null));
    // $FlowIssue
    fetchRatesData.mockReturnValue(
      Promise.resolve({
        currencies: {
          edges: [],
        },
      }),
    );
    // $FlowIssue
    getAvailableCurrencies.mockReturnValue(currencies);
    // $FlowIssue
    getRequestCurrency.mockReturnValue(null);
    // $FlowIssue
    resolveCurrency.mockReturnValue(currencies.eur);
    // $FlowIssue
    resolveRates.mockImplementation((data, available) => available);
  });

  afterEach(() => {
    // $FlowIssue
    fetchGeoData.mockReset();
    // $FlowIssue
    fetchRatesData.mockReset();
    // $FlowIssue
    getAvailableCurrencies.mockReset();
    // $FlowIssue
    getRequestCurrency.mockReset();
    // $FlowIssue
    resolveCurrency.mockReset();
    // $FlowIssue
    resolveRates.mockReset();
    // $FlowIssue
    store.getValue.mockReset();
    // $FlowIssue
    store.saveValue.mockReset();
  });

  test("render with router", () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={[{ pathname: "/", key: "testKey" }]}>
        <Provider all={currencies} countries={countries} fromLanguage="">
          Content
        </Provider>
      </MemoryRouter>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("render", () => {
    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("available", () => {
    const mockAvailable = {
      huf: {
        id: "huf",
        name: "Hungarian Forint",
        format: "__price__ Ft",
        uncertainFormat: false,
        round: "0",
        enabledOnAffilId: "",
        fallback: "",
        rate: 0.00312772,
      },
    };

    // $FlowIssue
    getAvailableCurrencies.mockReturnValue(mockAvailable);

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "affilid=TEST_AFFIL" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    expect(getAvailableCurrencies).toBeCalledWith(currencies, "TEST_AFFIL");
    expect(wrapper.prop("value").available).toEqual(mockAvailable);
  });

  test("current - with requested currency", () => {
    const mockCurrent = {
      huf: {
        id: "huf",
        name: "Hungarian Forint",
        format: "__price__ Ft",
        uncertainFormat: false,
        round: "0",
        enabledOnAffilId: "",
        fallback: "",
        rate: 0.00312772,
      },
    };

    // $FlowIssue
    resolveCurrency.mockReturnValue(mockCurrent);

    // $FlowIssue
    getRequestCurrency.mockReturnValue("REQ");

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage="LNG"
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    wrapper.setState({ selected: "SEL" });

    wrapper.update();

    expect(resolveCurrency).toBeCalledWith(currencies, currencies, ["SEL", "REQ", null, "LNG"]);
    expect(fetchGeoData).not.toBeCalled();
    expect(wrapper.prop("value").current).toEqual(mockCurrent);
  });

  test("current - by country", async () => {
    const promise = Promise.resolve({
      geoIP: {
        isoCountryCode: "cz",
      },
    });

    // $FlowIssue
    fetchGeoData.mockReturnValue(promise);

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage="LNG"
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    wrapper.setState({ selected: "SEL" });

    await promise;

    wrapper.update();

    expect(resolveCurrency).toBeCalledWith(currencies, currencies, ["SEL", null, "czk", "LNG"]);
    expect(fetchGeoData).toBeCalled();
  });

  test("current - by country - country unknown", async () => {
    const promise = Promise.resolve({
      geoIP: {
        isoCountryCode: "test",
      },
    });

    // $FlowIssue
    fetchGeoData.mockReturnValue(promise);

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage="LNG"
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    wrapper.setState({ selected: "SEL" });

    await promise;

    wrapper.update();

    expect(resolveCurrency).toBeCalledWith(currencies, currencies, ["SEL", null, null, "LNG"]);
    expect(fetchGeoData).toBeCalled();
  });

  test("current - by country - fetch failed", async () => {
    const promise = Promise.reject();

    // $FlowIssue
    fetchGeoData.mockReturnValue(promise);

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    await promise.catch(() => null);

    wrapper.update();

    expect(resolveCurrency).toBeCalledWith(currencies, currencies, [null, null, null, ""]);
    expect(fetchGeoData).toBeCalled();
  });

  test("setCurrency", () => {
    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    wrapper.instance().setCurrency("czk");

    wrapper.update();

    expect(resolveCurrency).toBeCalledWith(currencies, currencies, ["czk", null, null, ""]);
    expect(wrapper.state("selected")).toBe("czk");
    expect(store.saveValue).toBeCalledWith("czk");
  });

  test("setCurrency - unknown", () => {
    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    wrapper.instance().setCurrency("TST");

    wrapper.update();

    expect(resolveCurrency).toBeCalledWith(currencies, currencies, [null, null, null, ""]);
    expect(wrapper.state("selected")).toBe(null);
    expect(store.saveValue).not.toBeCalledWith();
  });

  test("rates", async () => {
    const mockRates = {
      currencies: {
        edges: [{ code: "czk", rate: 2 }],
      },
    };

    const promise = Promise.resolve(mockRates);

    // $FlowIssue
    fetchRatesData.mockReturnValue(promise);

    const mockResolved = {
      czk: {
        id: "czk",
        name: "Czech Republic Koruna",
        format: "__price__ Kč",
        uncertainFormat: false,
        round: "0",
        enabledOnAffilId: "",
        fallback: "",
        rate: 2,
      },
    };

    // $FlowIssue
    resolveRates.mockReturnValue(mockResolved);

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    await promise;

    wrapper.update();

    expect(fetchRatesData).toBeCalled();
    expect(resolveRates).toBeCalledWith(mockRates, currencies);
    expect(wrapper.prop("value").available).toEqual(mockResolved);
  });

  test("rates - fetch dailed", async () => {
    const promise = Promise.reject();

    // $FlowIssue
    fetchRatesData.mockReturnValue(promise);

    const wrapper = shallow(
      // $FlowIssue
      <Provider.WrappedComponent
        all={currencies}
        countries={countries}
        fromLanguage=""
        location={{ search: "" }}
      >
        Content
      </Provider.WrappedComponent>,
    );

    await promise.catch(() => null);

    wrapper.update();

    expect(fetchRatesData).toBeCalled();
    expect(resolveRates).not.toBeCalledWith();
    expect(wrapper.prop("value").available).toEqual(currencies);
  });
});
