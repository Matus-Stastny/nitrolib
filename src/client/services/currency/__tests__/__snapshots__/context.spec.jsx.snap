// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#currency/context render 1`] = `"Content"`;
